### ISSUE TITLE
[RFC] Tweaking the Release Schedule 

### ISSUE THREAD

**Issue by GawainLynch** [bolt/bolt#5239]
**NOTE:** This is a follow up to #2985 from last year.

Please add comments here, and I have added this as a discussion item for 

### Background

Over the last 3.5 years, we've built a great community and a core team that manages to all get along well, both personally and professionally.

The original agreed concept was to run something along the lines of

| Window   | Open (09:00 UTC) | Close (midnight UTC) |
|----------|------------------|----------------------|
| Feature  | 09 March 2014    | 22 March 2014        |
| Bug Fix  | 23 March 2014    | 05 April 2014        |
| Release  | 06 April 2014    | n/a                  |

Which translated to:
- 14 day feature window (hard limit)
- 14 day bug fix window (soft end-limit)
- 14 day contingency window (only if required)
- Release!

For a number of reasons the 2.3-dev cycle didn't run that way, and over a year later we have a tired, grumpy, and stressed team. 

This needs to change :wink:

### 3.0.0 Release

What will be released in early May as 3.0.0 has changed significantly from 2.x. For that reason, and to rest the core team members — restoring peace and harmony to the Bolt Universe — I suggest that we hold off opening the feature window 16 May (with the possible exception of the initial Bootstrap stuff for the admin UI from @bobdenotter and @rarila)

### 3.1 and beyond

I would like to get us to short, tight, & fast minor releases. We have a lot to do in the 3.x life cycle, and having long branches again is not going to help us do a good job of this.

#### 3.1 Release Schedule

Assuming we agree on 16 May as the point we reopen the feature window for 3.x, I propose that the 3.1 release schedule aims for:

| Window   | Open (09:00 UTC) | Close (midnight UTC) |
|----------|------------------|----------------------|
| Feature  | 16 May 2016      | 29 May 2016          |
| Bug Fix  | 30 May 2016      | 12 June 2016         |
| Release  | 13 June 2016     | n/a                  |

#### 3.2 Release Schedule

Assuming we close the 3.1 feature window on 30 May, we would then branch `release/3.1` but continue to allow feature work in `master` that would land in 3.2, making that schedule look similar to: 

| Window   | Open (09:00 UTC) | Close (midnight UTC) |
|----------|------------------|----------------------|
| Feature  | 30 May 2016      | 12 June 2016         |
| Bug Fix  | 13 June 2016     | 26 June 2016         |
| Release  | 27 June 2016     | n/a                  |

### Addressing Concerns

Historically there has been some concern that feature windows distract/prevent rapid, and focused, bug fix outcomes on major/minor pre-release branches. 

However, I would argue that we're hampering our own progression in terms of project output, and that this approach contributed to the root cause factors in the time blow out that was `2.3-dev`/`3.0-dev`.

To illustrate why I don't think this concern is valid… Between the 2.2.0 tag and the HEAD of `release/3.0` (d75284e3663e87534b824d796f77e1736c5de02c) the diff stat is currently:
```
1954 files changed, 108670 insertions(+), 59127 deletions(-)
```
Looking at the diff stats between 2.2.x tags (and release dates), there has been no shortage of work on keeping bugs under control:
```
 619 files changed, 19661 insertions(+), 9532 deletions(-)
```

With the break down being
- 2.2.0  — 2015-06-03
- 2.2.1  — 2015-06-05 — 8 files changed, 101 insertions(+), 36 deletions(-)
- 2.2.2  — 2015-06-12 — 11 files changed, 882 insertions(+), 39 deletions(-)
- 2.2.3  — 2015-06-15 — 5 files changed, 41 insertions(+), 99 deletions(-)
- 2.2.4  — 2015-06-25 — 11 files changed, 119 insertions(+), 33 deletions(-)
- 2.2.5  — 2015-07-24 — 36 files changed, 473 insertions(+), 211 deletions(-)
- 2.2.6  — 2015-07-28 — 35 files changed, 704 insertions(+), 420 deletions(-)
- 2.2.7  — 2015-07-29 — 3 files changed, 4 insertions(+), 8 deletions(-)
- 2.2.8  — 2015-07-31 — 4 files changed, 57 insertions(+), 13 deletions(-)
- 2.2.9  — 2015-08-28 — 377 files changed, 5297 insertions(+), 3467 deletions(-)
- 2.2.10 — 2015-09-01 — 6 files changed, 38 insertions(+), 4 deletions(-)
- 2.2.11 — 2015-09-21 — 51 files changed, 354 insertions(+), 154 deletions(-)
- 2.2.12 — Not released
- 2.2.13 — 2015-10-07 — 64 files changed, 2883 insertions(+), 1577 deletions(-)
- 2.2.14 — 2015-11-26 — 190 files changed, 5237 insertions(+), 343 deletions(-)
- 2.2.15 — 2015-12-29 — 21 files changed, 131 insertions(+), 39 deletions(-)
- 2.2.16 — 2016-01-01 — 23 files changed, 1002 insertions(+), 968 deletions(-)
- 2.2.17 — 2016-01-26 — 124 files changed, 535 insertions(+), 318 deletions(-)
- 2.2.18 — 2016-02-09 — 18 files changed, 95 insertions(+), 55 deletions(-)
- 2.2.19 — 2016-03-11 — 218 files changed, 1274 insertions(+), 5478 deletions(-)
- 2.2.20 — 2016-04-19 — 96 files changed, 10248 insertions(+), 6084 deletions(-)

…put succinctly 167,797 lines changed in development branches with nearly 30,00 lines still being changed in the stable branch during the same period… I think we're up to handling this! :smile:

**Comment by bobdenotter** [comment]
Overall, good stuff! Wholeheartedly agree! 👍 

> (with the possible exception of the initial Bootstrap stuff for the admin UI from @bobdenotter and @rarila)

No need to make an exception as far as I am concerned. This is going to take a short while, so no rush to get it in quicker. 

> Release	13 June 2016
> Release	27 June 2016

Just two weeks between 3.1 and 3.2? Isn't that a bit _too_ rapid? :-)

**Comment by GawainLynch** [comment]
> Just two weeks between 3.1 and 3.2? Isn't that a bit too rapid? 

Well, no. It is almost impossible for that to be an outcome as the bug fix window has to run its course and the whole approach will need refinement over time. More the goal here is that there is always features able to land in master, but we branch pre-releases in short spurts to minimise the amount of bug fix work that actually needs to happen.

**Comment by GawainLynch** [comment]
Oh, and I think I might have highlighted the wrong point above with "always", that is just a side effect, I think the aim is just to have smaller and faster minor release branches… released every 4-6 weeks.

**Comment by GawainLynch** [comment]
I've had a couple of private conversations with people and want to convey some points from those here:

* The minor version release cycle, per-release, should be approximately 4 to 6 weeks
* `master` should (almost) always be open for features
* Core team members would only respond to bugs affecting supported versions, i.e. current LTS and stable trees
* Bug fixes to LTS would cascade down to stable & master, e.g. if the current stable was 3.7 and current LTS was 3.3:
  * A bug fix to 3.3 would indirectly cascade down, via merge, to 3.4, 3.5 & 3.6
  * A bug report that *only* existed in 3.4 would be fixed by upgrading to 3.7 (i.e. latest stable)
  * PRs for said bug to 3.4 would be rejected
* The LoC delta & upgrade risk between versions like 3.4.x and 3.7.x would, practically speaking, be roughly similar to the differences between (for example) 2.2.13 and 2.2.19 — with the obvious exception of features

**Comment by GawainLynch** [comment]
Gist of above agreed upon at last night's meeting… Let's make it so!
